### PR TITLE
bpo-28180: Implementation for PEP 538

### PR DESCRIPTION
**Pull request by ncoghlan**
Reference implementation for PEP 538, which is in turn a partial fix for bpo-28180.

This updates the CPython CLI to attempt to set `LC_CTYPE` to a suitable UTF-8
based locale before loading the runtime when it detects that it is running in
the C locale.

It also updates the CPython runtime to emit a compatibility warning on `stderr`
when running in the C locale.

Remaining work:

- [x] Fix the Windows compatibility issue reported by Appveyor
- [x] What's New entry
- [x] Add a note in the What's New entry regarding the current locale coercion warning that indicates we're actively seeking feedback on it during the 3.7 pre-release cycle. Our experience with the Fedora 26 backport so far has been that the warning is useful in spotting places where we should probably change the configured locale to C.UTF-8 (e.g. in RPM build environments), but I'm still starting to wonder if it might be better for us to disable it before the F26 final release.
- [x] NEWS entry

### CODE REVIEW

File: Programs/python.c [hunk 1]
```diff
@@ -15,6 +15,110 @@ wmain(int argc, wchar_t **argv)
...
+        fprintf(stderr, _C_LOCALE_COERCION_WARNING, "LC_CTYPE", newloc);
+    } else {
+        fprintf(stderr, "Locale coercion must target LC_ALL or LC_CTYPE\n");
+        return;
+    }
+
+    /* Set PYTHONIOENCODING if not already set */
+    if (setenv("PYTHONIOENCODING", "utf-8:surrogateescape", 0)) {
+        fprintf(stderr,
+                "Error setting PYTHONIOENCODING during C locale coercion\n");
```
Comment (methane) [inline Programs/python.c]: This may break old Python 2 in subprocess.

```
$ cat t.py
# encoding: utf-8
print(u"こんにちは")

$ /usr/bin/python2.6 t.py
こんにちは

$ export PYTHONIOENCODING=utf-8:surrogateescape
$ /usr/bin/python2.6 t.py
TypeError: an integer is required
```

Comment (methane) [reply Programs/python.c]: I'm sorry, my environment was wrong.
I accidently test above in `Lib/` directory in cpython's checkout.

Comment (ncoghlan) [reply Programs/python.c]: It still raises a good question though, as that setting *does* affect Python 2 differently from the way it affects Python 3 - it changes the implicit encoding step on stdout, but stdin still relies on passing the raw bytes through without interpretation:

```
$ LANG=C python2
Python 2.7.13 (default, Jan 12 2017, 17:59:37) 
[GCC 6.3.1 20161221 (Red Hat 6.3.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> print(u"こんにちは")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-14: ordinal not in range(128)
>>> print("こんにちは")
こんにちは
>>> 
$ PYTHONIOENCODING=utf-8:surrogateescape LANG=C python2
Python 2.7.13 (default, Jan 12 2017, 17:59:37) 
[GCC 6.3.1 20161221 (Red Hat 6.3.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> print(u"こんにちは")
こんにちは
>>> print("こんにちは")
こんにちは
>>> 
```

It's also a potential problem that 'surrogateescape' doesn't exist in Python 2, so it may be better to just use `Py_SetStandardStreamEncoding` in PEP 538, and leave enabling `surrogateescape` in subprocesses as well to PEP 540 (via `PYTHONUTF8=1` in the parent environment).

It also turns out that `LANG=C python2` is an easy way to demonstrate that GNU readline just plain doesn't handle UTF-8 properly in the C locale - attempting to edit the `print(u"こんにちは")` line at the interactive prompt to remove the `u` prefix or add it back results in nonsense:

```
>>> print(u"こんにちは")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-14: ordinal not in range(128)
>>> print(�こんにちは")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-13: ordinal not in range(128)
>>> print("こんにちは")
こんにちは
>>> print(u��にちは") ")
こ�u��にちは
```

File: Doc/using/cmdline.rst [hunk 2]
```diff
@@ -713,6 +713,35 @@ conflict.
...
    .. versionadded:: 3.6
 
+
+.. envvar:: PYTHONCOERCECLOCALE
+
+   If set to a non-empty string, causes the main Python command line application
+   to skip coercing the legacy ASCII-based C locale to a more capable UTF-8
+   based alternative. Note that this setting is checked even when the
+   :option:`-E` or :option:`-I` options are used, as it is handled prior to
+   the processing of command line options.
```
Comment (warsaw) [inline Doc/using/cmdline.rst]: Am I reading this right?  It seems odd that setting `PYTHONCOERCECLOCALE` would *disable* coercing the legacy locale.  The sense is exactly opposite of what I'd expect.  

Should the envar be named `PYTHONNOCOERCECLOCALE`?

Also, what if the environment variable is set to `"0"`?  Given the above description, that should still skip coercion.

File: Lib/test/test_locale_coercion.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+def _set_locale_in_subprocess(locale_name, category):
+    cmd_fmt = "import locale; print(locale.setlocale(locale.{}, '{}'))"
+    cmd = cmd_fmt.format(category, locale_name)
+    result, py_cmd = run_python_until_end("-c", cmd, __isolated=True)
+    return result.rc == 0
+
+# Details of the CLI warning emitted at runtime
+CLI_COERCION_WARNING_FMT = (
+    "Python detected LC_CTYPE=C: {} coerced to {} (set another locale "
+    "or PYTHONCOERCECLOCALE=0 to disable this locale coercion behaviour)."
```
Comment (warsaw) [inline Lib/test/test_locale_coercion.py]: Does the value of the envar matter or not?  It just needs to be any string value, right?

File: Lib/test/test_locale_coercion.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+
+    def test_C_utf8_locale(self):
+        # Ensure the C.UTF-8 locale is accepted entirely without complaint
+        base_var_dict = {
+            "LANG": "",
+            "LC_CTYPE": "",
+            "LC_ALL": "",
+        }
+        for env_var in ("LC_ALL", "LC_CTYPE", "LANG"):
```
Comment (warsaw) [inline Lib/test/test_locale_coercion.py]: Does the order matter?  If not, what about using `for env_var in base_var_dict:`?

File: Lib/test/test_locale_coercion.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        expected_warning = []
+        if coerce_c_locale != "0":
+            # Check C locale is coerced with a warning on stderr
+            expected_warning.append(self.EXPECTED_COERCION_WARNING)
+        base_var_dict = {
+            "LANG": "",
+            "LC_CTYPE": "",
+            "LC_ALL": "",
+        }
+        for env_var in ("LC_ALL", "LC_CTYPE", "LANG"):
```
Comment (warsaw) [inline Lib/test/test_locale_coercion.py]: And here...

File: Lib/test/test_locale_coercion.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                    self.assertEqual(fsencoding, expected_fsencoding)
+                    self.assertEqual(stderr_lines, expected_warning)
+
+
+    def test_test_PYTHONCOERCECLOCALE_not_set(self):
+        # This should coerce to the C.UTF-8 locale by default
+        self._check_c_locale_coercion("utf-8", coerce_c_locale=None)
+
+    def test_PYTHONCOERCECLOCALE_not_zero(self):
+        # *Any* string other that "0" is considered "set" for our purposes
```
Comment (warsaw) [inline Lib/test/test_locale_coercion.py]: Doesn't this contradict the documentation at the top of this PR?

File: Lib/test/test_locale_coercion.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        if not os.path.exists(exe):
+            self.skipTest("%r doesn't exist" % exe)
+        # This is needed otherwise we get a fatal error:
+        # "Py_Initialize: Unable to get the locale encoding
+        # LookupError: no codec search functions registered: can't find encoding"
+        self.oldcwd = os.getcwd()
+        os.chdir(basepath)
+
+    def tearDown(self):
+        os.chdir(self.oldcwd)
```
Comment (warsaw) [inline Lib/test/test_locale_coercion.py]: What about using `self.addCleanup(os.chdir, self.oldcwd)` in the `setUp()` instead and getting rid of the `tearDown()`?

File: Lib/test/test_locale_coercion.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+    def run_embedded_interpreter(self, *args):
+        """Runs a test in the embedded interpreter"""
+        cmd = [self.test_exe]
+        cmd.extend(args)
+        p = subprocess.Popen(cmd,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE,
+                             universal_newlines=True)
```
Comment (warsaw) [inline Lib/test/test_locale_coercion.py]: Does it make any sense to run this without `universal_newlines` also?  Since the locale changes should only affect text mode, that additional test would just ensure that nothing's changed when running subprocesses in bytes mode.

File: Programs/python.c [hunk 9]
```diff
@@ -15,6 +15,110 @@ wmain(int argc, wchar_t **argv)
 }
 #else
 
+/* Helpers to better handle the legacy C locale
```
Comment (warsaw) [inline Programs/python.c]: There's been some discussion about the "legacy C locale" nomenclature.  What about "bare C locale"?

File: Programs/python.c [hunk 10]
```diff
@@ -15,6 +15,110 @@ wmain(int argc, wchar_t **argv)
...
+ * more capable UTF-8 based alternative.
+ *
+ * See the documentation of the PYTHONCOERCECLOCALE setting for more details.
+ *
+ */
+
+#ifdef PY_COERCE_C_LOCALE
+static const char *_C_LOCALE_COERCION_WARNING =
+    "Python detected LC_CTYPE=C: %.20s coerced to %.20s (set another locale "
+    "or PYTHONCOERCECLOCALE=0 to disable this locale coercion behaviour).\n";
```
Comment (warsaw) [inline Programs/python.c]: This seems to come up in every project I work on.  Do we use UK or US spelling? :)

I think in Python we use US spellings.

File: Python/pylifecycle.c [hunk 11]
```diff
@@ -302,6 +302,31 @@ import_init(PyInterpreterState *interp, PyObject *sysmod)
...
+static void
+_emit_stderr_warning_for_c_locale(void)
+{
+    const char *coerce_c_locale = getenv("PYTHONCOERCECLOCALE");
+    /* We don't emit a warning if locale coercion has been explicitly disabled.
+     *
+     * For consistency with the corresponding check in Programs/python.c
+     * we ignore the Python -E and -I flags here.
+     */
+    if (coerce_c_locale == NULL || strncmp(coerce_c_locale, "0", 2) != 0) {
```
Comment (warsaw) [inline Python/pylifecycle.c]: You do this same check in two places.  Would it make sense to refactor that into a call or macro so there's no chance in getting out of sync?  (See my question at the top of the PR.)

**Review by warsaw — COMMENTED** [review]
Leaving out for the moment whether this is a good idea or not (we'll use the PEP process for that), I have some questions about the code.

File: Doc/using/cmdline.rst (hunk 2 above)
Comment (ncoghlan) [reply Doc/using/cmdline.rst]: Oops, that's a holdover from when the setting was PYTHONALLOWCLOCALE - presumably I changed the title of the section, but then got distracted by something else before updating the body.

File: Lib/test/test_locale_coercion.py (hunk 3 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: Docs bug - the warning matches the behaviour, the docs don't.

File: Lib/test/test_locale_coercion.py (hunk 7 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: Good point - I stole the basic structure from `test_capi` (simplified a bit since this doesn't need to run on Windows), so it's missing some modern niceties. With `addCleanup`, we wouldn't even need `self.oldcwd` as a record, we can just add the cleanup operation before changing the directory:

    self.addCleanup(os.chdir, os.getcwd())
    os.chdir(basepath)

File: Lib/test/test_locale_coercion.py (hunk 8 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: For this particular test, I don't think so, as it's just checking that the runtime warning gets emitted when we bypass the locale coercion in the standard Python CLI, rather than doing anything with the standard streams.

However, it *does* make sense for `test_locale_coercion` itself (which I should rename to `test_c_locale_coercion`): that should check the behaviour when the Python subprocess is run with the `-u` option for unbuffered binary streams.

File: Programs/python.c (hunk 9 above)
Comment (ncoghlan) [reply Programs/python.c]: Even glibc devs consider 7-bit ASCII a legacy encoding at this point, it's just an enormous ecosystem to try to migrate to a new default way of doing things.

My view would likely be different if we were planning to invest time and energy in getting Python 3 to "work properly" in the C locale, but the underlying assumptions are sufficiently incompatible that it isn't even clear what "working properly" would even mean. (PEP 540 gets closer than the status quo, but it's definition of "working properly" is "ignoring the declared locale encoding as almost certainly being wrong")

File: Programs/python.c (hunk 10 above)
Comment (ncoghlan) [reply Programs/python.c]: Yeah, I'll typically use my native spellings when writing PEPs and email, but anything in the actual code or docs should be using the US spelling.

File: Python/pylifecycle.c (hunk 11 above)
Comment (ncoghlan) [reply Python/pylifecycle.c]: An alternative would be to check a C global here, rather than checking the environment variable directly, and have a `_Py_DisableCLocaleWarning()` helper that the CLI calls when `PYTHONCOERCECLOCALE=0` is set.

File: Doc/using/cmdline.rst (hunk 2 above)
Comment (ncoghlan) [reply Doc/using/cmdline.rst]: Fixed in https://github.com/python/cpython/pull/659/commits/1c3a2706ef276ca66f676218395b9d792ef3ca0e

File: Lib/test/test_locale_coercion.py (hunk 3 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: Docs fixed in https://github.com/python/cpython/pull/659/commits/1c3a2706ef276ca66f676218395b9d792ef3ca0e

File: Lib/test/test_locale_coercion.py (hunk 6 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: Docs fixed in https://github.com/python/cpython/pull/659/commits/1c3a2706ef276ca66f676218395b9d792ef3ca0e

File: Lib/test/test_locale_coercion.py (hunk 4 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: Done in https://github.com/python/cpython/pull/659/commits/d12b4126779b41c945c210124ba23b1c25ab62f5

File: Lib/test/test_locale_coercion.py (hunk 5 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: Done in https://github.com/python/cpython/pull/659/commits/d12b4126779b41c945c210124ba23b1c25ab62f5

File: Lib/test/test_locale_coercion.py (hunk 7 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: Done in https://github.com/python/cpython/pull/659/commits/d12b4126779b41c945c210124ba23b1c25ab62f5

File: Lib/test/test_locale_coercion.py (hunk 8 above)
Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: So it turns out that even `-u` doesn't change anything of interest to this PEP any more, and there's just an error in the `--help` docs that makes it sound like it still does: http://bugs.python.org/issue28647

Comment (ncoghlan) [reply Lib/test/test_locale_coercion.py]: However, now that the PEP potentially also affects the way the standard streams are configured, I expanded on the test cases to also cover that: https://github.com/python/cpython/pull/659/commits/4e6d5029601a74b12807f5e776964cd062c2194c

File: Python/pylifecycle.c (hunk 11 above)
Comment (ncoghlan) [reply Python/pylifecycle.c]: The downside of that approach would be that it means embedding applications can't use `PYTHONCOERCECLOCALE=0` to silence the warning without changing the configured locale, and I don't want to define a new public embedding configuration API just to avoid checking the environment variable twice.

So I'll unconditionally export a private helper function from pylifecycle.c (so the ABI doesn't depend on the configure flag setting), and re-use that from the CLI implementation.

File: Programs/python.c (hunk 10 above)
Comment (ncoghlan) [reply Programs/python.c]: Fixed in https://github.com/python/cpython/pull/659/commits/ccfc83f04a3b7d5ac46a82dc4836e2f7f46702ee

**Comment by ncoghlan** [comment]
@zooba @brettcannon Would one of you have some time to look at the Windows failure here? I suspect what's happened is that I've changed some code that I thought was *nix specific, and it needs a preprocessor guard to get it to retain the old behaviour on Windows.

File: Doc/using/cmdline.rst [hunk 12]
```diff
@@ -713,6 +713,40 @@ conflict.
...
+
+   If this variable is *not* set, or is set to a value other than ``0``, and
+   the current locale reported for the ``LC_CTYPE`` category is the default
+   ``C`` locale, then the Python CLI will attempt to configure one of the
+   following locales for the given locale categories before loading the
+   interpreter runtime:
+
+   * ``C.UTF-8`` (``LC_ALL``)
+   * ``C.utf8`` (``LC_ALL``)
+   * ``UTF-8`` (``LC_CTYPE``)
```
Comment (methane) [inline Doc/using/cmdline.rst]: These `LC_ALL` should be updated.

**Comment by zooba** [comment]
@ncoghlan Is there some state involved in checking/configuring the locale override that _must_ be run before `Py_Initialize`? On Windows, `main.c` is exactly one line long and you haven't updated it to call the new functions, so it may be related to that. I don't see any changes to common code that could be a problem, but maybe I missed it.

As an aside, I'm somewhat uncomfortable about the "let's ignore -E -S" parts, but if you're sure it's not going to enable loading arbitrary code (via an encoding) then I won't push back too hard.

File: Python/pylifecycle.c [hunk 13]
```diff
@@ -327,6 +328,157 @@ initexternalimport(PyInterpreterState *interp)
...
+};
+
+static char *
+get_default_standard_stream_error_handler(void)
+{
+    const char *ctype_loc = setlocale(LC_CTYPE, NULL);
+    if (ctype_loc != NULL) {
+        /* "surrogateescape" is the default in the legacy C locale */
+        if (strcmp(ctype_loc, "C") == 0) {
+            return "surrogateescape";
```
Comment (ncoghlan) [inline Python/pylifecycle.c:385]: @zooba This is the bit that I suspect may be unintentionally changing the interpreter startup behaviour on Windows and thus causing the Appveyor test failures: there's currently nothing restricting this particular code path to only run on non-Windows systems, so I suspect it may be triggering and activating `surrogateescape` by default on the Windows standard streams, even though Windows uses its own code page system, rather than the POSIX locale system. If I'm right, then the entire body of this function should be inside some kind of `#ifdef`, such that the Windows case degenerates to `return NULL;`

File: Python/pylifecycle.c [hunk 14]
```diff
@@ -327,6 +328,157 @@ initexternalimport(PyInterpreterState *interp)
...
+_Py_CoerceLegacyLocale(void)
+{
+#ifdef PY_COERCE_C_LOCALE
+    /* We ignore the Python -E and -I flags here, as the CLI needs to sort out
+     * the locale settings *before* we try to do anything with the command
+     * line arguments. For cross-platform debugging purposes, we also need
+     * to give end users a way to force even scripts that are otherwise
+     * isolated from their environment to use the legacy ASCII-centric C
+     * locale.
+    */
```
Comment (ncoghlan) [inline Python/pylifecycle.c]: Per @zooba: I need to explain the (lack of) security implications here, since:

- the envvar can only be used to turn off locale coercion, it doesn't do anything else (i.e. it's a boolean toggle, not input data)
- if an attacker is in a position to corrupt the behaviour of the `ascii` encoding, they're likely to be able to attack the `utf-8` encoding just as effectively

Comment (zooba) [reply Python/pylifecycle.c]: 👍

File: Python/pylifecycle.c (hunk 13 above)
Comment (zooba) [reply Python/pylifecycle.c:385]: Possibly. The only thing that is available for coercion on Windows is `locale.getpreferredencoding()`, and there are no results from `setlocale()` that you can reasonably interpret as "wrong".

When/if the "force UTF-8" option comes along, that should apply to Windows, but I don't think we need to determine the error handler here at all for Windows.

Comment (ncoghlan) [reply Python/pylifecycle.c:385]: Comparing this to the old code (as a result of the `test_sys` failure), it seems we *were* setting `surrogateescape` by default in the C locale, even on Windows. So I've had a second go at fixing this by keeping that, and only skipping the new checks for the coercion target locales.

Comment (ncoghlan) [reply Python/pylifecycle.c:385]: It turned out this wasn't the problem, but rather the earlier call to `setlocale(LC_CTYPE, "")`, where I had removed a `HAVE_SETLOCALE` guard as being redundant in 2017. It turns out that even though Windows does have `setlocale` available:

1. Passing an empty string doesn't work the same way it does in glibc (presumably it syncs the active C locale with the active Windows code page rather than looking at environment variables)
2. The Windows builds don't define `HAVE_SETLOCALE`, so that guard was previously serving as a cryptic `#ifndef MS_WINDOWS` check

So rather than restoring the original guard, I replaced it with an explicit `#ifndef MS_WINDOWS` check instead.

**Comment by ncoghlan** [comment]
Huzzah, finally sorted out all the subtle inconsistencies with the behaviour on Windows :)

@methane, @zooba, @warsaw Do any of you want to take a further look at the code changes while I finish up with the What's New entry and adding some caveats around the current locale coercion warning?

**Comment by ncoghlan** [comment]
The What's New entry has been added, so I consider this PR to be feature complete now - to reduce the chance of conflicts, I won't add the NEWS entry until we're just about to merge it.

**Comment by ncoghlan** [comment]
AppVeyor seems to be MIA, but the last Windows run passed, and I've only made docs and metadata changes since then, so I'll be going ahead with the merge and marking the PEP as Final once the Travis run completes.

File: Programs/python.c [hunk 15]
```diff
@@ -70,7 +103,11 @@ main(int argc, char **argv)
 
     /* Force again malloc() allocator to release memory blocks allocated
        before Py_Main() */
+#ifdef Py_DEBUG
```
Comment (ronaldoussoren) [inline Programs/python.c:106]: This appears to be an unrelated change.

Comment (ncoghlan) [reply Programs/python.c:106]: You're right that it's technically unrelated now, but an earlier iteration of the patch failed in debug mode without it - that iteration was using `Py_SetStandardStreamEncoding`, and when you do that, you allocate some memory that gets freed during interpreter startup *after* the default allocator has been configured. Without this change, that free attempt failed in debug mode due to the allocator mismatch.

So these guards mean that the "safe to call before `Py_Initialize`" APIs actually are safe to call from the CLI entry point.
